### PR TITLE
Extend MCP/ACP listing capabilties

### PR DESCRIPTION
**Pull request by avishniakov**
## Describe changes
I implemented a few improvements to the listing capabilities of MCP entities:
- `zenml model version list` reworked and no longer mandates model name (but supports it via param) + it supports listing with tag filters (via `-t` or `--tag` params)
- `zenml model list` supports listing with tag filters (via `-t` or `--tag` params)
- `zenml artifact list` supports listing with tag filters (via `-t` or `--tag` params)
- `zenml artifact version list` supports listing with tag filters (via `-t` or `--tag` params)
- Client methods `Client.list_models` and `Client.list_model_versions` also support filtering by tags
- Client methods `Client.list_artifacts` and `Client.list_artifact_versions` also support filtering by tags
- Minor performance improvement to CLI model version listing to avoid checking the DB state (bug OSSK-343)

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).
- [ ] If my changes require changes to the dashboard, these changes are communicated/requested.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

## Summary by CodeRabbit

- **New Features**
	- Enhanced model listing with optional tag filtering.
	- Improved model version listing with additional filter options for model name and tags.

- **Enhancements**
	- Renamed command for consistency: `model endpoint_artifacts` to `model deployment_artifacts`.

- **Bug Fixes**
	- Fixed model version listing to properly handle multiple tag filters.

- **Documentation**
	- Updated command-line interface documentation to reflect new flags and parameters.

- **Tests**
	- Added integration tests for tag-based filtering in model and model version listings.
	- Updated existing tests to accommodate new tag functionality.

### CODE REVIEW

**Comment by avishniakov** [comment]
@coderabbitai review

File: tests/integration/functional/test_client.py [hunk 1]
```diff
@@ -1323,6 +1323,26 @@ def test_latest_version_retrieval(self, clean_client: "Client"):
         assert model_.latest_version_name == mv2.name
         assert model_.latest_version_id == mv2.id
 
+    def test_list_by_tags(self, clean_client: "Client"):
```
Comment (strickvl) [inline tests/integration/functional/test_client.py:1326]: maybe add a test for the empty list passed in etc?

```python
ms = clean_client.list_models(tags=[])
# Expected behavior for no tags scenario

ms = clean_client.list_models(tags=[""])
# Expected behavior for empty string tag

ms = clean_client.list_models(tags=["non_existent_tag"])
assert len(ms) == 0
```

Comment (strickvl) [reply tests/integration/functional/test_client.py:1326]: and the same below for the others?

File: src/zenml/cli/artifact.py [hunk 2]
```diff
@@ -281,3 +296,28 @@ def prune_artifacts(
...
+        "type": artifact_version.type,
+        "materializer": artifact_version.materializer,
+        "data_type": artifact_version.data_type,
+        "tags": [t.name for t in artifact_version.tags],
+    }
+
+
+def _artifact_to_print(
+    artifact_version: ArtifactResponse,
+) -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/cli/artifact.py:318]: It is already `Union[str,UUID,List[str]]`. I can update this, but I don't care much cause it is printer utility for CLI. Keeping it as Any and letting `__repr__` do the rest is easier here.
For bigger functions like `_artifact_version_to_print` it will be a nigtmare 🙂  See 6+ types already.

File: src/zenml/cli/model.py [hunk 3]
```diff
@@ -77,21 +84,26 @@ def model() -> None:
...
+    "--tag",
+    "-t",
+    help="Tags to search for.",
     type=str,
     required=False,
+    multiple=True,
 )
 @model.command("list", help="List models with filter.")
-def list_models(name: str) -> None:
+def list_models(tag: Optional[List[str]], **kwargs: Any) -> None:
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/cli/model.py:97]: Mypy will go crazy, precisely it should be `Dict[str,Any]`, but normally people put `Any` for `*args,**kwargs` from what I see

File: tests/integration/functional/test_client.py (hunk 1 above)
Comment (avishniakov) [reply tests/integration/functional/test_client.py:1326]: Thanks, caught one bug 👍🏼

File: src/zenml/cli/model.py [hunk 4]
```diff
@@ -54,13 +55,19 @@ def _model_to_print(model: ModelResponse) -> Dict[str, Any]:
 def _model_version_to_print(
     model_version: ModelVersionResponse,
 ) -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/cli/model.py:57]: Same as above

**Review by strickvl — APPROVED** [review]
Looks good from my side now, assuming tests pass.
